arbornml/xmlwrap.hpp

```diff
@@ -0,0 +1,310 @@
+bool nl_from_cstr(non_negative& out, const char* content);
+bool nl_from_cstr(long long& out, const char* content);
+bool nl_from_cstr(double& out, const char* content);
+
+// Wrap xmlChar* NUL-terminated string that requires deallocation.
+
+struct xml_string {
+    explicit xml_string(const xmlChar* p): p_(p, xml_string::deleter) {}
+
+    operator const char*() const {
+        return reinterpret_cast<const char*>(p_.get());
+    }
+
+private:
+    std::shared_ptr<const xmlChar> p_;
+    static void deleter(const xmlChar* x) { xmlFree((void*)x); }
+};
+
+// Wrappers below are generally constructed with two arguments,
+// a pointer corresponding to the libxml2 object, and a dependency
+// object (typically shared_ptr<X> for some X) that guards the
+// lifetime of another object upon which this depends.
+
+template <typename XmlType>
+void trivial_dealloc(XmlType*) {}
+
+template <typename XmlType, void (*xml_dealloc)(XmlType *) = &trivial_dealloc<XmlType>>
+struct xml_base {
+    xml_base(XmlType* p, std::any depends = {}):
+        p_(p, xml_dealloc),
+        depends_(std::move(depends))
+    {}
+
+protected:
+    // Access to raw wrapped pointer type.
+    XmlType* get() const { return p_.get(); }
+
+    // Copy of shared_ptr<> governing lifetime of referenced object.
+    auto self() const { return p_; }
+
+    // Copy of dependency object.
+    std::any depends() const { return depends_; }
+
+private:
+    std::shared_ptr<XmlType> p_;
+    std::any depends_;
+};
+
+// xmlNode RAII wrapper (non-owning).
+
+struct xml_node: protected xml_base<xmlNode>  {
+    using base = xml_base<xmlNode>;
+    explicit xml_node(xmlNode* p, std::any depends):
+        base(p, std::move(depends))
+    {}
+
+    bool is_element() const { return get()->type==XML_ELEMENT_NODE; }
+    bool is_attr() const { return get()->type==XML_ATTRIBUTE_NODE; }
+    xml_string content() const { return xml_string(xmlNodeGetContent(get())); }
+    unsigned line() const { return get()->line; }
+
+    bool has_prop(const char* name) const { return xmlHasProp(get(), (const xmlChar*)name); }
+
+    template <typename T>
+    T prop(const char* name, std::optional<T> default_value = std::nullopt) const {
+        xmlChar* c = xmlGetProp(get(), (const xmlChar*)(name));
+        if (!c) {
+            return default_value? default_value.value(): throw parse_error("missing required attribute", get()->line);
+        }
+
+        T v;
+        return nl_from_cstr(v, reinterpret_cast<const char*>(c))? v: throw parse_error("attribute type error", get()->line);
+    }
+
+    using base::get; // (unsafe access)
+};
+
+// xmlNodeSet RAII wrapper; resource lifetime is governed by an xmlXPathObject.
+
+struct xml_nodeset: protected xml_base<xmlNodeSet>  {
+    using base = xml_base<xmlNodeSet>;
+
+    xml_nodeset(): xml_nodeset(nullptr, std::any{}) {}
+
+    xml_nodeset(xmlNodeSet* p, std::any depends):
+        base(p, std::move(depends))
+    {}
+
+    struct iterator {
```

Comment:
We could, except `xml_node` has no non-const methods, other than the unsafe `get`. And I think if you need that, then we can presume you know what you are doing.

Reply:
But this would allow you to modify without using the unsafe `get`? How about making it a const pointer?

Reply:
We'd then have to chuck `const_cast`s into the `xml_node` method implementations: libxml2 is not const-aware.

Reply:
All the wrappers are presenting an immutable interface, so we don't need to worry too much on that front; we could put a const_cast in `xml_base::get` though and return a const pointer from the iterator `operator->`

Reply:
Yep, it turns out it's entirely fine to return a const pointer. I suspect it didn't work at some point previously, because I wouldn't have put it in as a lark :)